### ISSUE TITLE
Reading application settings in NET WinForms OOP designer fails with exception 'Invalid URI: The hostname could not be parsed"

### ISSUE THREAD

**Issue by tracktownsoftware** [dotnet/runtime#70318]
### Environment

Version 17.2.2

### .NET version

net6.0-windows

### Did this work in a previous version of Visual Studio and/or previous .NET release?

No, this is new in VS2022 and the NET WinForms OOP designer.

### Issue description

This is an issue in the new NET WinForms designer when using the microsoft.winforms.designer.sdk to add design-time support to a custom NET6 WinForms control. Reading application settings at design-time fails with exception “Invalid URI: this hostname could not be parsed”.

I also reported this as feedback from VS2022:
https://developercommunity.visualstudio.com/t/Reading-application-settings-fails-in-th/10055004

### Steps to reproduce

My github sample to reproduce:
https://github.com/tracktownsoftware/MicrosoftWinFormsDesignerSDK_AppSettingsBug

The above repository includes a Nuget package you can use to test and reproduce. The nuget package contains a custom NET6 WinForms "MyButton" control with WinForms Designer Extensibility SDK design-time support.

**Comment by RussKie** [comment]
@jeffhandley @krwq putting this on your radar.

**Comment by buyaa-n** [comment]
Thanks for the clear repro @tracktownsoftware and the additional details @RussKie as you pointed out the issue caused by the AppDomain.CurrentDomain.BaseDirectory path prefixed with `\\?\` when creating Uri at:
https://github.com/dotnet/runtime/blob/c8654e7378a085c8221028cf0ebd759d800ceba7/src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ClientConfigPaths.cs#L251
I could be able to fix this by removing the prefix if exist or something like that, but it seems better to be fixed at the [Uri constructor](https://github.com/dotnet/runtime/issues/58712) or [AppDomain.CurrentDomain.BaseDirectory level](https://github.com/dotnet/runtime/issues/58714) as @jkotas suggested [in previous PR](https://github.com/dotnet/runtime/pull/58627#issuecomment-912630223) so that we will not see similar issues again. One of the issues even closed, so I would suggest you to go to those issues and ask for urgent fix

**Comment by RussKie** [comment]
@karelz this is a core scenario for the Windows Forms designer

**Comment by danmoseley** [comment]
@MihaZupan

**Comment by buyaa-n** [comment]
Another workaround could be move the Uri creation code 
https://github.com/dotnet/runtime/blob/c8654e7378a085c8221028cf0ebd759d800ceba7/src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ClientConfigPaths.cs#L251
to the location where it is used: https://github.com/dotnet/runtime/blob/c8654e7378a085c8221028cf0ebd759d800ceba7/src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ClientConfigPaths.cs#L269
This workaround at least works for the repro attached

**Comment by karelz** [comment]
I just answered on the other issue - https://github.com/dotnet/runtime/issues/58712#issuecomment-1161543088, trying to explain the context and limitations we are facing in the `Uri` space.
If there is reasonable workaround, we should strongly consider it before investing into `Uri` feature which is risky this late in the release cycle.

Who can help us better understand the details of the above scenario of `\\?\` paths in Uri? Mainly why and how it is used, why `Uri` instead of `Path`, etc.

**Comment by jkotas** [comment]
I agree with @karelz that fixing anything in Uri is complex and has very high risks. It does not make sense to push to address this in Uri for .NET 7. It would be a new feature and it may require new APIs.

There are cheaper and lower risk options to address the specific WinForms issue. We should look at those: Workaround it in the System.Configuration.ConfigurationManager library itself, or fix https://github.com/dotnet/runtime/issues/58714

**Comment by jkotas** [comment]
> the details of the above scenario of \\?\ paths in Uri? Mainly why and how it is used, why Uri instead of Path, etc.

Compatibility with .NET Framework. 

The original .NET Framework code is here: https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.Configuration/System/Configuration/ClientConfigPaths.cs#L343-L411 . Notice that it does not use the problematic Uri constructor.

This is trying to mimic .NET Framework logic that was based on Code Access Security. It looks like that this logic deviated over time from .NET Framework one through series of bug fixes, so it does not fully match what .NET Framework did anyway. Another option to address the WinForms issues is to delete the part of the logic that tries to use Uris. It may be an observable behavior change, but I doubt that anybody will notice.

**Comment by RussKie** [comment]
Looks like the discussion has split between here and https://github.com/dotnet/runtime/issues/58712. And since the latter is closed, in attempt to merge the forks, @karelz I'm going to reply to your questions here:

> Key questions I have: How does it work on .NET Framework? Is it a breaking change between .NET Core and .NET Framework?

In .NET Framework the Windows Forms designer is run in the VS process (as both the host and the designer use the same runtime). To facilitate the .NET support the designer was moved out-of-proc, and the server process is started with `\\?\` path.



> Are there workaround options?

There aren't any workarounds we can do in the designer AFAIK - as this is purely a .NET runtime scenario. I first reported at almost a year ago in https://github.com/dotnet/runtime/issues/56897.
As for the .NET runtime - looks like @buyaa-n and @jkotas may have found a way to fix it, https://github.com/dotnet/runtime/issues/70318#issuecomment-1162160391.

It is worth noting, that the Windows Forms out-of-process designer codebase is targeting .NET Core 3.1 so that our customers can design .NET Windows Forms apps targeting .NET Core 3.1 and above. With .NET Core 3.1 starting to near its EOL there's no appetite to get this fixed in it, but we'd want the fix to be serviced in to .NET 6, as this bug is blocking the adoption (e.g., this one or https://github.com/dotnet/runtime/issues/56897).

**Comment by buyaa-n** [comment]
> but we'd want the fix to be serviced in to .NET 6, as this bug is blocking the adoption (e.g., this one or https://github.com/dotnet/runtime/issues/56897)

Tagging @ericstj @jeffhandley for considering porting to .NET 6,

**Comment by RussKie** [comment]
Thank you @buyaa-n for the fix. I'll give it a try once it flows through to dotnet/winforms.

**Comment by ericstj** [comment]
Can folks familiar with the bug and fix consider the servicing bug bar to see if they'd like to make the case to fix in .NET 6.0 servicing?  If so, then the next steps would be to prepare such a PR and apply the template.

**Comment by buyaa-n** [comment]
> Can folks familiar with the bug and fix consider the servicing bug bar to see if they'd like to make the case to fix in .NET 6.0 servicing? If so, then the next steps would be to prepare such a PR and apply the template.

In our area pod triage we discussed this and decided to move forward with servicing request, the bot created the porting PR but I just found that 6.0 branch doesn't have [the previous fix](https://github.com/dotnet/runtime/pull/58627) @krwq did, without that the fix is incomplete, not sure what we do in such case, I could cherry-pick the @krwq's change into the PR

**Comment by RussKie** [comment]
Could you give me a hand and point me a release that I can test? I've downloaded the latest nightly (7.0.100-preview.7.22329.1) from dotnet/installer, and it only contains the runtime build from 15 June (7.0.0-preview.6.22314.11+eeb0c1551b5fc9ccd6ceeaf5d2b1504afab249d7), and the fix was merged on 23 June.
Thank you

**Comment by jkotas** [comment]
Try https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md#daily-builds-table

**Comment by RussKie** [comment]
I am afraid this won't work. The daily build contains only Microsoft.NETCore.App workload, and the fix went into System.Configuration.ConfigurationManager.dll, which is a part of Microsoft.WindowsDesktop.App workload.

**Comment by jkotas** [comment]
You should be able to get daily build of the System.Configuration.ConfigurationManager.dll package using the instructions at the top of the document.

**Comment by buyaa-n** [comment]
> Could you give me a hand and point me a release that I can test?

In addition to what @jkotas suggestion I also sent you email with `System.Configuration.ConfigurationManager.6.0.1.nupkg` package built from 6.0 branch with the fixes

**Comment by buyaa-n** [comment]
@RussKie please let us know your test result so that we can complete the [servicing request](https://github.com/dotnet/runtime/pull/71452)

**Comment by RussKie** [comment]
Thank you @buyaa-n.

I installed 7.0.100-rc.1.22371.1 available at https://github.com/dotnet/installer, and tested the repro provided in https://github.com/tracktownsoftware/MicrosoftWinFormsDesignerSDK_AppSettingsBug. I created a Windows Forms app targeting .NET 7 and referenced the built NuGet from the repro project.

The issue is fixed:

https://user-images.githubusercontent.com/4403806/180335038-680632d9-507d-4653-9f5e-cd4bb7577c4d.mp4
